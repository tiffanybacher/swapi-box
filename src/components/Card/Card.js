import React, { Component } from 'react';
import { getSubData } from '../../apiCalls';

class Card extends Component {
  state = {
    category: this.props.category,
    name: this.props.data.name
  }

  getSpeciesData() {
    const { data } = this.props;

    return getSubData(data.species)
      .then(data => this.setState({
        species: data.name
      }))
      .catch(error => this.setState({
        errorStatus: 'Error adding species data'
      }));
  }

  getHomeworldData() {
    const { data } = this.props;

    return getSubData(data.homeworld)
      .then(data => this.setState({ 
          homeworld: {
            name: data.name, 
            population: data.population
          } 
      }))
      .catch(error => this.setSate({
        errorStatus: 'Error adding homeworld data'
      }));
  }

  getResidentsData() {
    const { data } = this.props;

    data.residents.forEach(residentURL => {
      return getSubData(residentURL)
        .then(resident => {
          let residents;

          if (!this.state.residents) {
            residents = [resident.name];
          } else {
            residents = [...this.state.residents, resident.name];
          }

          this.setState({ residents });
        })
        .catch(error => this.setState({
          errorStatus: 'Failed to load resident data'
        }));
    });

  }

  render() {
    let subData;

    if (this.state.category === 'people') {
      this.getSpeciesData();
      this.getHomeworldData();
      
      const { species } = this.state || '';
      const { name } = this.state.homeworld || '';
      const { population } = this.state.homeworld || '';

      subData = 
        <div className="subData">
          <h3>Species: {species}</h3>
          <h3>Homeworld: {name}</h3>
          <h3>Population: {population}</h3>
          <button>Save</button>
        </div>
    } else if (this.state.category === 'planets') {
      // this.getResidentsData();
    
      // const { residents } = this.state || [];
      // const residentList = residents.map(resident => {
        // return <p>{resident}</p>
      const { population, terrain, climate } = this.props.data;

      subData = 
        <div className="subData">
          <h3>Population: {population}</h3>
          <h3>Terrain: {terrain}</h3>
          <h3>Climate: {climate}</h3>
          {/*<h3>Residents: {}</h3>*/}
        </div>
    }

    return (
      <article className='Card'>
        <h2>{this.state.name}</h2>
        {subData}
      </article>
    );
  }
}

export default Card;