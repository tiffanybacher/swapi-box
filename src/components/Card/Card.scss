.Card {
  border: 3px solid black;
  border-radius: 3px;
  margin: 20px;
  padding: 20px;
  width: 200px;
} 