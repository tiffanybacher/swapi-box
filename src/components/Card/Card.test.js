import React from 'react';
import { shallow } from 'enzyme';
import Card from './Card';

describe('Card', () => {
  let mockCategory = 'people';
  let mockData = {
          "name": "Luke Skywalker",
          "height": "172",
          "mass": "77",
          "hair_color": "blond",
          "skin_color": "fair",
          "eye_color": "blue",
          "birth_year": "19BBY",
          "gender": "male",
          "homeworld": "https://swapi.co/api/planets/1/",
          "films": [
              "https://swapi.co/api/films/2/",
              "https://swapi.co/api/films/6/",
              "https://swapi.co/api/films/3/",
              "https://swapi.co/api/films/1/",
              "https://swapi.co/api/films/7/"
          ],
          "species": [
              "https://swapi.co/api/species/1/"
          ],
          "vehicles": [
              "https://swapi.co/api/vehicles/14/",
              "https://swapi.co/api/vehicles/30/"
          ],
          "starships": [
              "https://swapi.co/api/starships/12/",
              "https://swapi.co/api/starships/22/"
          ],
          "created": "2014-12-09T13:50:51.644000Z",
          "edited": "2014-12-20T21:17:56.891000Z",
          "url": "https://swapi.co/api/people/1/"
        }
  let mockPlanetData = {
    "name": "Coruscant",
    "rotation_period": "24",
    "orbital_period": "368",
    "diameter": "12240",
    "climate": "temperate",
    "gravity": "1 standard",
    "terrain": "cityscape, mountains",
    "surface_water": "unknown",
    "population": "1000000000000",
    "residents": [
        "https://swapi.co/api/people/34/",
        "https://swapi.co/api/people/55/",
        "https://swapi.co/api/people/74/"
    ],
    "films": [
        "https://swapi.co/api/films/5/",
        "https://swapi.co/api/films/4/",
        "https://swapi.co/api/films/6/",
        "https://swapi.co/api/films/3/"
    ],
    "created": "2014-12-10T11:54:13.921000Z",
    "edited": "2014-12-20T20:58:18.432000Z",
    "url": "https://swapi.co/api/planets/9/"
  }
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Card 
        category={mockCategory} 
        data={mockData} 
        key={mockData.created}
      />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({
      category: 'people',
      name: 'Luke Skywalker'
    });
  });

  describe('getSpeciesData', () => {
    it('should update state with species', async () => {
      await wrapper.instance().getSpeciesData();

      expect(wrapper.state('species')).toEqual('Human');
    });
  });

  describe('getHomeworldData', () => {
    it('should update state with homeworld name and population', async () => {
      await wrapper.instance().getHomeworldData();

      expect(wrapper.state('homeworld')).toEqual('Tatooine');

      expect(wrapper.state('population')).toEqual('200000');
    });
  });

  describe('getResidentsData', () => {
    it.skip('should set state with all the resident names', async () => {
      mockData = mockPlanetData;

      await wrapper.instance().getResidentsData();

      expect(wrapper.state('residents')).toEqual(['Finis Valorum', 'Adi Gallia', 'Jocasta Nu']);
    });
  });
});