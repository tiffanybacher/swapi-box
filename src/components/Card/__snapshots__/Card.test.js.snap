// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<article
  className="Card"
>
  <h2>
    Luke Skywalker
  </h2>
  <div
    className="subData"
  >
    <h3>
      Species: 
    </h3>
    <h3>
      Homeworld: 
    </h3>
    <h3>
      Population: 
    </h3>
    <button>
      Save
    </button>
  </div>
</article>
`;
